pkg/protoassert/assert.go

```diff
@@ -1,130 +1,194 @@
+	return true
 }
 
-func SliceContains[V any, T message[V]](t testing.TB, slice []T, element T, msgAndArgs ...interface{}) bool {
+// SliceContains mimics [assert.Contains].
+func SliceContains[S ~[]T, T message[V], V any](t testing.TB, s S, contains T, msgAndArgs ...any) bool {
 	t.Helper()
-	for _, e := range slice {
-		if proto.Equal(e, element) {
+	for _, e := range s {
+		// Do not use [Equal] here, as it will unnecessarily log unequal elements.
+		if e.EqualVT(contains) {
 			return true
 		}
 	}
-	return assert.Failf(t, "Slice does not contain element", "%q %v", element.String(), msgAndArgs)
+	return assert.Fail(t, fmt.Sprintf("%#v does not contain %#v", s, contains), msgAndArgs...)
 }
 
-func SliceNotContains[V any, T message[V]](t testing.TB, slice []T, element T, msgAndArgs ...interface{}) bool {
+// SliceNotContains mimics [assert.NotContains].
+func SliceNotContains[S ~[]T, T message[V], V any](t testing.TB, s S, contains T, msgAndArgs ...any) bool {
 	t.Helper()
-	for _, e := range slice {
-		if proto.Equal(e, element) {
-			return assert.Failf(t, "Slice contain element", "%q %v", element.String(), msgAndArgs)
+	for _, e := range s {
+		// Do not use [Equal] here, as it will unnecessarily log unequal elements.
+		if e.EqualVT(contains) {
+			return assert.Fail(t, fmt.Sprintf("%#v should not contain %#v", s, contains), msgAndArgs...)
 		}
 	}
 	return true
 }
 
-func ElementsMatch[V any, T message[V]](t testing.TB, expected, actual []T, msgAndArgs ...interface{}) bool {
+// ElementsMatch mimics [assert.ElementsMatch].
+func ElementsMatch[S ~[]T, T message[V], V any](t testing.TB, expected, actual S, msgAndArgs ...any) bool {
 	t.Helper()
-	areEqual := assert.Len(t, actual, len(expected))
-	for _, e := range expected {
-		areEqual = SliceContains(t, actual, e) && areEqual
-	}
-	for _, a := range actual {
-		areEqual = SliceContains(t, expected, a) && areEqual
+	if len(expected) == 0 && len(actual) == 0 {
+		return true
 	}
-	if !areEqual {
-		t.Log(msgAndArgs...)
+	extraExpected, extraActual := diffSlices(expected, actual)
+	if len(extraExpected) == 0 && len(extraActual) == 0 {
+		return true
 	}
-	return areEqual
+	return assert.Fail(t, formatSliceDiff(expected, actual, extraExpected, extraActual), msgAndArgs...)
 }
 
-func MapSliceEqual[K comparable, V any, T message[V]](t testing.TB, expected, actual map[K][]T, msgAndArgs ...interface{}) bool {
+// MapSliceEqual determines if the expected and actual maps (from key to slice) are equal.
+func MapSliceEqual[M ~map[K][]T, K comparable, T message[V], V any](t testing.TB, expected, actual M, msgAndArgs ...any) bool {
 	t.Helper()
 	expectedKeys := maps.Keys(expected)
 	actualKeys := maps.Keys(actual)
-	areEqual := !assert.ElementsMatch(t, expectedKeys, actualKeys)
-	for expectedKey, expectedValue := range expected {
-		a := actual[expectedKey]
-		areEqual = SlicesEqual(t, expectedValue, a, expectedKey) && areEqual
+	if !assert.ElementsMatch(t, expectedKeys, actualKeys, msgAndArgs...) {
```

Comment:
nitpick: Can we prepend information that the keys are different because now we get only `elements differ`, but it's not clear what elements are representing.